File: ruby/complex-numbers/complex_numbers.rb
# NOTE: Many of these operations are provided by Ruby's CMath library

class ComplexNumber
  attr_reader :real, :imaginary

  def initialize(a, b)
    @real = a
    @imaginary = b
  end

  def +(other)
    a = @real + other.real
    b = @imaginary + other.imaginary
    return ComplexNumber.new(a, b)
  end

  def -(other)
    a = @real - other.real
    b = @imaginary - other.imaginary
    return ComplexNumber.new(a, b)
  end

  def *(other)
    a = @real * other.real - @imaginary * other.imaginary
    b = @imaginary * other.real + @real * other.imaginary
    return ComplexNumber.new(a, b)
  end

  def /(other)
    a = (@real * other.real + @imaginary * other.imaginary)
      .fdiv(other.real ** 2 + other.imaginary ** 2)
    b = (@imaginary * other.real - @real * other.imaginary)
      .fdiv(other.real ** 2 + other.imaginary ** 2)
    return ComplexNumber.new(a, b)
  end

  def ==(other)
    @real == other.real && @imaginary == other.imaginary
  end

  def conjugate
    ComplexNumber.new(@real, -@imaginary)
  end

  def abs
    Math.sqrt(@real ** 2 + @imaginary ** 2)
  end

  def exp
    if @imaginary == Math::PI
      a = -(Math::E ** @real).round
      b = 0
    else
      a = Math::E ** @real * Math.cos(@imaginary)
      b = Math::E ** @real * Math.sin(@imaginary)
    end

    return ComplexNumber.new(a, b)
  end
end

File: javascript/complex-numbers/complex-numbers.js
export class ComplexNumber {
    constructor(a, b) {
        this.real = a;
        this.imag = b;
    }

    add(other) {
        let a = this.real + other.real;
        let b = this.imag + other.imag;
        return new ComplexNumber(a, b);
    }

    sub(other) {
        let a = this.real - other.real;
        let b = this.imag - other.imag;
        return new ComplexNumber(a, b);
    }

    div(other) {
        let a = (this.real * other.real + this.imag * other.imag) /
            (other.real ** 2 + other.imag ** 2);
        let b = (this.imag * other.real - this.real * other.imag) /
            (other.real ** 2 + other.imag ** 2);
        return new ComplexNumber(a, b);
    }

    mul(other) {
        let a = this.real * other.real - this.imag * other.imag;
        let b = this.imag * other.real + this.real * other.imag;
        return new ComplexNumber(a, b);
    }

    get abs() {
        return Math.sqrt(this.real ** 2 + this.imag ** 2);
    }

    get conj() {
        let a = this.real;
        let b;

        if (this.imag == 0)
            b = 0;
        else
            b = -this.imag;

        return new ComplexNumber(a, b);
    }

    get exp() {
        let a = Math.E ** this.real * Math.cos(this.imag);
        let b = Math.E ** this.real * Math.sin(this.imag);

        return new ComplexNumber(a, b);
    }
}
